flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/collect/CollectSinkFunctionRandomITCase.java

```diff
@@ -0,0 +1,366 @@
+
+/**
+ * Random IT cases for {@link CollectSinkFunction}.
+ * It will perform random insert, random checkpoint and random restart.
+ */
+public class CollectSinkFunctionRandomITCase extends TestLogger {
+
+	private static final int MAX_RESULTS_PER_BATCH = 3;
+	private static final JobID TEST_JOB_ID = new JobID();
+	private static final OperatorID TEST_OPERATOR_ID = new OperatorID();
+
+	private static final TypeSerializer<Integer> serializer = IntSerializer.INSTANCE;
+
+	private CollectSinkFunctionTestWrapper<Integer> functionWrapper;
+	private boolean jobFinished;
+
+	@Test
+	public void testUncheckpointedFunction() throws Exception {
+		// run multiple times for this random test
+		for (int testCount = 30; testCount > 0; testCount--) {
+			functionWrapper = new CollectSinkFunctionTestWrapper<>(serializer, MAX_RESULTS_PER_BATCH * 4);
+			jobFinished = false;
+
+			List<Integer> expected = new ArrayList<>();
+			for (int i = 0; i < 50; i++) {
+				expected.add(i);
+			}
+			Thread feeder = new ThreadWithException(new UncheckpointedDataFeeder(expected));
+
+			List<Integer> actual = runFunctionRandomTest(feeder);
+			assertResultsEqualAfterSort(expected, actual);
+
+			functionWrapper.closeWrapper();
+		}
+	}
+
+	@Test
+	public void testCheckpointedFunction() throws Exception {
+		// run multiple times for this random test
+		for (int testCount = 30; testCount > 0; testCount--) {
+			functionWrapper = new CollectSinkFunctionTestWrapper<>(serializer, MAX_RESULTS_PER_BATCH * 4);
+			jobFinished = false;
+
+			List<Integer> expected = new ArrayList<>();
+			for (int i = 0; i < 50; i++) {
+				expected.add(i);
+			}
+			Thread feeder = new ThreadWithException(new CheckpointedDataFeeder(expected));
+
+			List<Integer> actual = runFunctionRandomTest(feeder);
+			assertResultsEqualAfterSort(expected, actual);
+
+			functionWrapper.closeWrapper();
+		}
+	}
+
+	private List<Integer> runFunctionRandomTest(Thread feeder) throws Exception {
+		CollectClient collectClient = new CollectClient();
+		Thread client = new ThreadWithException(collectClient);
+
+		Thread.UncaughtExceptionHandler exceptionHandler = (t, e) -> {
+			feeder.interrupt();
+			client.interrupt();
+			e.printStackTrace();
+		};
+		feeder.setUncaughtExceptionHandler(exceptionHandler);
+		client.setUncaughtExceptionHandler(exceptionHandler);
+
+		feeder.start();
+		client.start();
+		feeder.join();
+		client.join();
+
+		return collectClient.results;
+	}
+
+	private void assertResultsEqualAfterSort(List<Integer> expected, List<Integer> actual) {
+		Collections.sort(expected);
+		Collections.sort(actual);
+		Assert.assertThat(actual, CoreMatchers.is(expected));
+	}
+
+	/**
+	 * A {@link RunnableWithException} feeding data to the function. It will fail when half of the data is fed.
+	 */
+	private class UncheckpointedDataFeeder implements RunnableWithException {
+
+		private LinkedList<Integer> data;
+		private final List<Integer> checkpointedData;
```

Comment:
`checkpointedData` for  `UncheckpointedDataFeeder` ?

Reply:
`originalData` might be the proper name.